clang/lib/CodeGen/ItaniumCXXABI.cpp

```diff
@@ -3565,8 +3565,32 @@ void ItaniumRTTIBuilder::BuildVTablePointer(const Type *Ty) {
   // Check if the alias exists. If it doesn't, then get or create the global.
   if (CGM.getItaniumVTableContext().isRelativeLayout())
     VTable = CGM.getModule().getNamedAlias(VTableName);
-  if (!VTable)
-    VTable = CGM.getModule().getOrInsertGlobal(VTableName, CGM.Int8PtrTy);
+
+  // To generate valid device code global pointers should have global address
+  // space in SYCL.
+  bool GenTyInfoGVWithGlobalAS =
+      CGM.getLangOpts().SYCLIsDevice &&
+      CGM.getLangOpts().SYCLAllowVirtualFunctions &&
+      (VTableName == ClassTypeInfo || VTableName == SIClassTypeInfo);
+  auto VTableTy = GenTyInfoGVWithGlobalAS
+                      ? CGM.Int8Ty->getPointerTo(
+                            static_cast<unsigned>(LangAS::opencl_global))
+                      : CGM.Int8PtrTy;
+  if (!VTable) {
+    if (GenTyInfoGVWithGlobalAS) {
+      VTable = CGM.getModule().getOrInsertGlobal(VTableName, VTableTy, [&] {
+        return new llvm::GlobalVariable(
+            CGM.getModule(), VTableTy, /*isConstant=*/false,
+            llvm::GlobalVariable::ExternalLinkage, /*Initializer=*/nullptr,
+            VTableName, /*InsertBefore=*/nullptr,
+            llvm::GlobalValue::ThreadLocalMode::NotThreadLocal,
+            llvm::Optional<unsigned>(
+                static_cast<unsigned>(LangAS::opencl_global)));
```

Comment:
```suggestion
                getContext().getTargetAddressSpace(LangAS::opencl_global)));
```

Reply:
Still opencl_global?

Reply:
is there a difference ? :trollface: 

I guess this is more pedantic in this context

```suggestion
                static_cast<unsigned>(LangAS::sycl_global)));
```

Reply:
There should be no difference, but @KornevNikita says that these are different. https://github.com/intel/llvm/pull/7255#discussion_r1013829973.

I think with `getContext().getTargetAddressSpace()` there will be no difference for SYCL to SPIR compilation, because SPIR address space map uses the same target address space value for both language address spaces.

Reply:
The static cast is a misuse of the name address space in clang, the numerical value they have has no meaning as such. As it depends on the target, you must use the address space map.

> is there a difference ? :trollface:

I was gently trolling you :) I have no intention on reopening that discussion.

Reply:
Done [e5ab819](https://github.com/intel/llvm/pull/7255/commits/e5ab819233ba2ae80afff51b67367b7649071da9)